modules/mqtt/testcontainers/mosquitto/__init__.py

```diff
@@ -0,0 +1,247 @@
+            self.client.on_message = MosquittoContainer.on_message
+            self.client.loop_start()  # launch a thread to call loop() and dequeue the message
+
+        return self.client, err
+
+    def start(self) -> Self:
+        super().start()
+        self._connect()
+        return self
+
+    def stop(self, force=True, delete_volume=True) -> None:
+        self.client.disconnect()
+        self.client = None  # force recreation of the client object at next start()
+        self.watched_topics = {}  # clean all watched topics as well
+        super().stop(force, delete_volume)
+
+    class WatchedTopicInfo:
+        def __init__(self):
+            self.count = 0
+            self.timestamp_start_watch = time.time()
+            self.last_payload = ""
+
+        def on_message(self, msg: mqtt_client.MQTTMessage):
+            self.count += 1
+            # for simplicity: assume strings are used in integration tests and are UTF8-encoded:
+            self.last_payload = msg.payload.decode("UTF-8")
+
+        def get_count(self):
+            return self.count
+
+        def get_last_payload(self):
+            return self.last_payload
+
+        def get_rate(self):
+            duration = time.time() - self.timestamp_start_watch
+            if duration > 0:
+                return self.count / duration
+            return 0
+
+    def on_message(client: mqtt_client.Client, mosquitto_container: "MosquittoContainer", msg: mqtt_client.MQTTMessage):
+        # very verbose but useful for debug:
+        # print(f"Received `{msg.payload.decode()}` from `{msg.topic}` topic")
+        if msg.topic == "$SYS/broker/messages/received":
+            mosquitto_container.msg_queue.put(msg)
+        else:
+            # this should be a topic added through the watch_topics() API...
+            # just check it has not been removed (e.g. by unwatch_all):
+            if msg.topic in mosquitto_container.watched_topics:
+                mosquitto_container.watched_topics[msg.topic].on_message(msg)
+            else:
+                print(f"Received msg on topic [{msg.topic}] that is not being watched")
+
+    def get_messages_received(self) -> int:
+        """
+        Returns the total number of messages received by the broker so far.
+        """
+
+        client, err = self.get_client()
+        if not client.is_connected():
+            raise RuntimeError(f"Could not connect to Mosquitto broker: {err}")
+
+        client.subscribe("$SYS/broker/messages/received")
+
+        # wait till we get the first message from the topic;
+        # this wait will be up to 'sys_interval' second long (see mosquitto.conf)
+        try:
+            message = self.msg_queue.get(block=True, timeout=5)
+            return int(message.payload.decode())
+        except Queue.Empty:
+            return 0
+
+    def watch_topics(self, topics: list):
+        client, err = self.get_client()
```

Comment:
same here, seems to be doing too much

Reply:
right now this is an opt-in feature: if the user wants to "watch" a topic, he can use the watch_topics() API. If it's not used, there's no (performance) impact during the test...
so perhaps it's a "nice to have" that can stay?

Reply:
removed